### PR TITLE
Enable a new algorithm for choosing SPSA hyperparameters.

### PR DESCRIPTION
**Pull request by vdbergh**
This PR adds a checkbox to the "New Test" form. Checking this option will rewrite the hyperparameters furnished by the tuning code in Stockfish in such a way that the SPSA tune will finish within 0.5 Elo from the optimum (with 95% probability) assuming the function 'parameters->Elo' is quadratic and varies 2 Elo per parameter over each specified parameter interval. The hyperparameters are relatively conservative and their performance will degrade gracefully if the actual variation is different.  The theoretical basis for choosing the hyperparameters is given in this document
    
https://github.com/vdbergh/spsa_simul/blob/master/doc/theoretical_basis.pdf
    
The formulas can be checked by simulation which is done here
    
https://github.com/vdbergh/spsa_simul
    
It is important for the current version of the algorithm to be able to estimate the draw ratio. The PR uses a heuristic method to do so which makes various assumptions: book='noob_3moves.epd', contempt=24, threads=1 and no nodestime.

### CODE REVIEW

**Comment by vondele** [comment]
@vdbergh IMO, this should come with a few lines of description, for example the guidelines on picking the interval, and the underlying assumptions. Basically, to help the user to do the right thing without having the read the long thread, the simulator, etc... 

As one review comment 'New algorithm' is almost always a poor name. Do we have something more descriptive, and better in the longer term? (robust, quadratic, guaranteed, ... !? )

**Comment by vondele** [comment]
Minor nit 'the tuning branch' is now outdated, since the code is in master. What about 'the tuning routines in Stockfish'? 

Otherwise, this looks good, and presumably ready for some more in-depth testing... @ppigazzini should we have a look at this on the dev server?

**Comment by ppigazzini** [comment]
DEV updated.

**Comment by ppigazzini** [comment]
@vdbergh with 12 parameters it triggers `Number of games must be <= 800000`

**Comment by vdbergh** [comment]
> @vdbergh with 12 parameters it triggers `Number of games must be <= 800000`

There are some restrictions on the form values in the current js code which I did not touch since I do not know where they come from. @tomtor ?

**Comment by vondele** [comment]
I think this is a check that was put in place to avoid mistakes. However, it might make sense to relax things?

I think the form has a good 'feel'. 'SPSA A' parameter isn't changing, I don't know enough what it does to judge if that's right or wrong.

**Comment by vdbergh** [comment]
@vondele The value of A does not matter if alpha is zero but I guess psychologically it would be better to set it to zero also.  Also alpha and gamma should actually be mathematically zero but I did not test yet if this does not cause some server issue (I don't think it will).

**Comment by ppigazzini** [comment]
@vondele with a normal SPSA the "A" value (suggested 5-10% of total iterations) avoids big values for "ak" at the start of the optimization `ak = a / (k + 1 + A)^alpha`

@vdbergh I will try to set by hand 0 values to alpha and gamma

**Comment by ppigazzini** [comment]
Found a SPSA with only 4 parameters to make some tests :)

https://dfts-0.pigazzini.it/tests/view/5f02ff19919f102ab14f68b3

**Comment by tomtor** [comment]
> > @vdbergh with 12 parameters it triggers `Number of games must be <= 800000`
> 
> There are some restrictions on the form values in the current js code which I did not touch since I do not know where they come from. @tomtor ?

Yes, the 800000 limit is in place to guard against very large chunk collections. 

It was introduced because a user increased the number of games by just adding '00' to the end of the default and it crashed fishtest. Could probably be raised somewhat further, but we do need to restrict it. Do we really want/need (tuning) tests with > 800000 games? In the past there were somewhat successful runs with 100 parameters, but allocating the number of games as computed by this PR (10000000?) for 100 parameters is not possible anyway.

**Comment by vondele** [comment]
@tomtor, indeed makes sense still to keep an upper limit, and collect some experience with tuning a smaller number of parameters first. Maybe it just needs an additional line of info in the info-box that right now, the maximal number of parameters that can be tuned is approx. N since we limit the number of games per test to 800k. If it turns out we're highly successful with the new approach we can start thinking about how to expand it to more parameters.

**Comment by vdbergh** [comment]
@ppigazzini It is currently not possible to set alpha=gamma=0 manually since things are reset on submit. In any case this is not terribly important as `580000**0.001=1.013` (so very close to 1). 

The second and the third test are doing exactly the same thing. This is ok if we want to check that things are repeatable (which is always a very good idea), but I am not sure if there are enough resources are available to run 1.7 million games in a reasonable time frame, even if they are 1+0.01. I can currently only contribute 3 cores as the server I used to use can no longer run the latest version of cutechess-cli.

**Comment by ppigazzini** [comment]
@vdbergh I restarted the server on DEV after changing by hand `tests_run.mak` here:
```
504                 $("input[name='spsa_alpha']").val(0.00);
505                 $("input[name='spsa_gamma']").val(0.00);
```

This code already allows 0 values
```
218       <label class="field-label rightmost">SPSA Alpha</label>
219       <input type="number" min="0" step="0.001" name="spsa_alpha"
220              class="third-size no-arrows"
221              value="${args.get('spsa', {'alpha': '0.602'})['alpha']}" />
222
223       <label class="field-label" style="margin-left: 7px">SPSA Gamma</label>
224       <input type="number" min="0" step="0.001" name="spsa_gamma"
225              class="third-size no-arrows"
226              value="${args.get('spsa', {'gamma': '0.101'})['gamma']}" />
```

**Comment by ppigazzini** [comment]
> The second and the third test are doing exactly the same thing. This is ok if we want to check that things are repeatable (which is always a very good idea), but I am not sure if there are enough resources available to run 1.7 million games in a reasonable time frame, even if they are 1+0.01.

My idea is to check the PR code on DEV submitting some SPSA tests.
We could check the convergence on PROD pasting the parameters value by hand and selecting a SPSA test with good parameters, eg

https://dfts-0.pigazzini.it/tests/view/5f0331d41289d603ccb2ebd4


>  I can currently only contribute 3 cores as the server I used to use can no longer run the latest version of cutechess-cli.

Let me know the distro.

**Comment by vdbergh** [comment]
CentOS 6. Yes very old... I need to upgrade it but it takes some planning but as the server is headless and in a room I do not have a key for it takes too much planning...

**Comment by vondele** [comment]
BTW, I think we should use the testing on DEV to establish the implementation is behaving as expected. Testing if this new implementation is really helpful we should do on PROD, just because the DEV setup doesn't have the resources. Doing a few tests on PROD with the parameters as obtained on DEV might be useful however.

**Comment by vondele** [comment]
So, started one test on PROD https://tests.stockfishchess.org/tests/view/5f03768359f6f035328944db

I started them from the current values... probably an alternative would be to start somewhat away from the optimum and see if one goes back.

**Comment by ppigazzini** [comment]
DEV updated, started some tests.

**Comment by vondele** [comment]
spsa_new_params
Just an update with intermediate results & comparison of the two running SPSA runs mentioned before. The runs starting with displaced values have roughly found master values, but those starting from master have moved a bit away.  I might do an intermediate SPRT run on the latter values to see if they are better.

**Comment by vondele** [comment]
spsa_new_params_final
attached image is the full SPSA runs. I would say all looks consistent. We find more or less the values of master back.

I tried the near final results in a SPRT test, but like the first one this was not showing a passing result, suggesting master is already optimal. https://tests.stockfishchess.org/tests/view/5f09ad1559f6f0353289488c

So, I think this more or less shows that the new scheme is doing fine. I think we can consider merging this.

**Comment by ppigazzini** [comment]
@vdbergh please open a new PR for the stochastic rounding. I would like:
 - some tests of stochastic rounding with an integer parameter in a little range (e.g. depth in 11-15) and other parameters. If the only good depth is 13, all the SPSA iterations with depth=12 or depth=14 IMO are wasted and in that case a bias could be a good thing to have.
 - a clean up of the unnecessary code from server/worker
 - a good commit message to make clear that SPSA works with real numbers, and that the stochastic rounding has:
   - primary goal to deal with Stockfish integer parameters (here a Stockfish parameter scaling is useless)
   - secondary goal is to round all parameters values for the cutechess-cli parameter input constrain, (here a good Stockfish parameter scaling avoids a precision loss). This could be dropped with a future cutechess-cli release able to deal with floating point parameters values.

**Comment by vdbergh** [comment]
@ppigazzini  Cutechess-cli accepting real numbers does not make any difference since then the rounding needs to happen inside SF (as SF uses fixed precision). 

You need to do the rounding somewhere and in my opinion it is best done outside SF since then it is generic and the user does not have think about it.

**Comment by ppigazzini** [comment]
@vdbergh so it's more complicated that I thought, please open a new PR to collect the information.

**Comment by vdbergh** [comment]
@ppigazzini I don't really see the point of opening a new PR for what is just one line of code representing common sense (as I said stochastic rounding is common in machine learning). 

I could add to the info window that the SPSA algorithm should not be used for parameters that can just take a few values and that there is also no point in artificially introducing scaling to try to achieve this (this actually has nothing to do with the new algorithm, it is even more true for the algorithm in master). 

For the new algorithm an interval of length 6 represents the bare minimum (this could be enforced). Then c is 1 so that the gradient descent actually has some room to play.

It is an interesting question what would be a good tuning algorithm for a parameter that can take only a few values.  A simple round robin tournament seems tempting but then one does not use the fact that nearby parameter values are expected to be correlated. Some form of Bayesian tuning seems to be the most natural method to exploit this.

**Comment by ppigazzini** [comment]
@vdbergh the previous PR seems to be tested OK by @vondele, so if this is fine with you I would like to merge it asap to have some feedback from the framework.

Stochastic rounding was here for years and nobody used it because bad documentation, please mind also that some fishtest contributors could lack Stockfish knowledge (for example me).

So I would like to  have a new PR for stochastic rounding:
 - to document why we need a rounding function and why we are choosing the stochastic rounding
 - to clean up the server/worker code. Merging now https://github.com/glinscott/fishtest/pull/719/commits/ede07e37025f5b7990d50355ce095f6575143c63 will leave some dead code (the "deterministic" rounding) and some unnecessary code (the double level rounding in worker code)

**Comment by vdbergh** [comment]
> @vdbergh the previous PR seems to be tested OK by @vondele, so if this is fine with you I would like to merge it asap to have some feedback from the framework.
> 
> Stochastic rounding was here for years and nobody used it because bad documentation, please mind also that some fishtest contributors could lack Stockfish knowledge (for example me).
> 
> So I would like to have a new PR for stochastic rounding:

Ok I guess I can do that. 

As I said it is hard to say that we "need" stochastic rounding. It is just that it avoids that one has to think about rounding. It is similar to repeated openings. It may be hard to prove that this is truly harmful. But if there are no repeated openings (or as few as possible) one simply does not have think about it.

> 
> * to document why we need a rounding function and why we are choosing the stochastic rounding
> * to clean up the server/worker code. Merging now [ede07e3](https://github.com/glinscott/fishtest/commit/ede07e37025f5b7990d50355ce095f6575143c63) will leave some dead code (the "deterministic" rounding) and some unnecessary code (the double level rounding in worker code)

**Comment by vdbergh** [comment]
@ppigazzini I reverted the last commit and squashed the PR.

**Comment by vdbergh** [comment]
Thanks! Now let's see....

**Comment by ppigazzini** [comment]
PROD updated, thank you @vdbergh :)

**Comment by tomtor** [comment]
@vdbergh Thanks, this is a very nice update to the fishtest SPSA tuning feature!

**Comment by ppigazzini** [comment]
@vdbergh in a corner case (e.g. integer not scaled parameter with narrow range, old SPSA algorithm) stochastic rounding could change the ordering of `round(p+c)` and `round(p-c)` (e.g when `N<p-c<p+c<N+1`). I done some math and the expected difference should be still `2c`, so the occasionally strange reversed derivative shouldn't be a problem (at least for only one parameter). Please double check this case at your convenient time :)
